### DISCUSSION TITLE
Can't start minio on the same port but different local ip

### DISCUSSION THREAD

**Discussion by gusev-p** [minio/minio#17817] [Q&A] [answered]
This command ends up with `address already in use`:
```
/usr/local/bin/minio server --address "127.52.56.1:9000" ./data

API: SYSTEM()
Time: 11:31:20 UTC 08/08/2023
Error: listen tcp 127.0.0.1:9000: bind: address already in use (*net.OpError)
       2: internal/logger/logger.go:258:logger.LogIf()
       1: cmd/signals.go:80:cmd.handleSignals()
```
The ip address 9000 is already listened on by another minio instance, but on different local IP. The following code can listen on "127.52.56.1:9000" without errors:
```
# echo-server.py

import socket

HOST = "127.96.128.1"  # Standard loopback interface address (localhost)
# HOST = "127.0.0.1"
PORT = 9000  # Port to listen on (non-privileged ports are > 1023)

with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
    s.bind((HOST, PORT))
    s.listen()
    conn, addr = s.accept()
    with conn:
        print(f"Connected by {addr}")
        while True:
            data = conn.recv(1024)
            if not data:
                break
            conn.sendall(data)
```
minio version:
```
$ minio --version
minio version RELEASE.2023-05-04T21-44-30Z (commit-id=5569acd95cbc8c7d3fac59cf2aba77baa1b9bbeb)
Runtime: go1.20.3 linux/amd64
License: GNU AGPLv3 <https://www.gnu.org/licenses/agpl-3.0.html>
Copyright: 2015-2023 MinIO, Inc.
```

Is this expected behavior or am I doing something wrong?

**Comment by harshavardhana** [comment, marked as answer]
Yes this is because we need to bind to localhost, please use a different port.

**Comment by nyh** [comment]
> Yes this is because we need to bind to localhost, please use a different port.

I think this issue should be reopened. It's a real bug that `--address` is given, but being ignored.

You said that you "need to bind to localhost". Why? You can bind a listening socket to any IP address that belongs to any network interface of the machine. Localhost (just an alias to 127.0.0.1) is just one of many different IP address that a typical machine can listen on. Even on the loopback interface, 127.0.0.1 is just an example address - any of the 1.6 million addresses in 127/8 - e.g., 127.1.2.3 is a perfectly valid address to listen on (try it!).

If you're curious why this use case is useful, well, the idea that a single machine can host multiple independent servers on multiple IP addresses. For example, you can have multiple HTTP servers, one listens on 127.1.2.3:80 and the other on 127.2.3.4:80 - without needing to pick a different port - both can be 80 - and this works beautifully. We expect exactly the same from minio - you should be able to run two minio servers each listening on a different IP address, with the same default choice of port number.  This setup is useful for testing where you are trying to simulate an entire cluster on one test machine.

**Reply by harshavardhana** [reply]
We need to bind to localhost always that is by design. So there is nothing to fix here.

Please use a different port.

**Comment by nyh** [comment]
The "--address" documentation says:

> Binds the [minio](https://min.io/docs/minio/linux/reference/minio-server/minio-server.html#command-minio.server) server process to a specific network address and port number. Specify the address and port as ADDRESS:PORT, where ADDRESS is an IP address or hostname and PORT is a valid and open port on the host system.

Where does it say that minio needs to "bind to localhost always"? And why does it, and how is that "by design"?
I still don't understand what "--address" does in this case - does it listen to the given address **and** also to localhost (127.0.0.1)?

And yes, using a different port is indeed a working workaround for running multiple minio servers on the same test machine. The ugly part is that it requires the client to know of this different port number, whereas the clients are currently aware of their different IP addresses, adding mapping to port numbers will be an additional complication that should have been unnecessary because there's usually no problem on Linux to run multiple servers listening to the same port on different IP addresses.

**Reply by harshavardhana** [reply]
As I said please use different port. If it works thats your answer..

I don't wish to engage beyond that.

**Reply by kbr-scylla** [reply]
Maybe the description of `--address` should be fixed?
